climt/_core/constants.py

```diff
@@ -0,0 +1,303 @@
+    'thermal_conductivity_of_liquid_phase': 'thermal_conductivity_of_liquid_water'
+}
+
+default_categories = dict(
+    planetary_constants=set(('gravitational_acceleration',
+                             'planetary_radius',
+                             'planetary_rotation_rate',
+                             'seconds_per_day')),
+
+    physical_constants=set(('stefan_boltzmann_constant',
+                            'avogadro_constant',
+                            'speed_of_light',
+                            'boltzmann_constant',
+                            'loschmidt_constant',
+                            'universal_gas_constant',
+                            'planck_constant')),
+
+    condensible_constants=set(('density_of_liquid_phase',
+                               'heat_capacity_of_liquid_phase',
+                               'heat_capacity_of_vapor_phase',
+                               'specific_enthalpy_of_vapor_phase',
+                               'gas_constant_of_vapor_phase',
+                               'latent_heat_of_condensation',
+                               'latent_heat_of_fusion',
+                               'density_of_solid_phase_as_ice',
+                               'density_of_solid_phase_as_snow',
+                               'heat_capacity_of_solid_phase_as_ice',
+                               'heat_capacity_of_solid_phase_as_snow',
+                               'thermal_conductivity_of_solid_phase_as_ice',
+                               'thermal_conductivity_of_solid_phase_as_snow',
+                               'thermal_conductivity_of_liquid_phase',
+                               'freezing_temperature_of_liquid_phase',)),
+
+    atmospheric_constants=set(('heat_capacity_of_dry_air_at_constant_pressure',
+                               'gas_constant_of_dry_air',
+                               'thermal_conductivity_of_dry_air',
+                               'reference_air_pressure')),
+
+    stellar_constants=set(('stellar_irradiance',)),
+
+    oceanographic_constants=set(()),
+
+    chemical_constants=set(('heat_capacity_of_water_vapor_at_constant_pressure',
+                            'density_of_liquid_water',
+                            'gas_constant_of_water_vapor',
+                            'latent_heat_of_vaporization_of_water',
+                            'heat_capacity_of_liquid_water',
+                            'latent_heat_of_fusion_of_water',)),
+
+    miscellaneous_constants=set(())
+)
+
+
+class ConstantLibrary(object):
+    """
+    Maintain constant values and categories. Currently available
+    constants are accessible as :code:`constant_library.current_constants`.
+    """
+
+    def __init__(self):
+
+        self.init_constant_library()
+
+    def init_constant_library(self):
+        """
+        Initialise the constant library from sympl's :code:`default_constants`.
+        """
+
+        self.current_constants = deepcopy(default_constants)
```

Comment:
You should always use Sympl's value for default_constants instead of a copy. If someone alters Sympl's default_constants directly, you want the CliMT components to use this updated value.

Reply:
If you want to use self.current_constants for readability, and you must use a map, you can write a custom object that will automatically apply the map when you do a lookup. But see map comment above.

Reply:
I agree that we should directly use default_constants instead of a copy. A copy should be kept and only used when reset_constants is called. Will fix this.

Reply:
We can't apply the map during runtime if the constant is modified. but the code I just pushed uses sympl's default_constants.